pilot/pkg/config/kube/ior/route.go

```diff
@@ -290,27 +182,91 @@ func (r *route) handleAdd(cfg config.Config) error {
 	r.gatewaysLock.Lock()
 	defer r.gatewaysLock.Unlock()
 
-	if _, ok := r.gatewaysMap[gatewaysMapKey(cfg.Namespace, cfg.Name)]; ok {
+	if _, ok := r.gatewayMap[gatewayMapKey(cfg.Namespace, cfg.Name)]; ok {
 		IORLog.Infof("gateway %s/%s already exists, not creating route(s) for it", cfg.Namespace, cfg.Name)
 		return nil
 	}
```

Comment:
In this function, we're handling the addition of a Gateway. Here, we check if the Gateway already exists. If this is indeed possible, this means that whatever code called this function assumes that an ADD event can only be received when the Gateway object is created. This code here indicates that that's not true. 

Also, if `onGatewayAdded` gets called and find that the Route for the Gateway already exists, it simply does nothing. What if the Route's configuration doesn't match the Gateway? IOR should update the Route if necessary. 

The same applies to the other `onGateway*` functions. These three functions should be a single function that updates the Route regardless of what caused the function to be invoked (either an ADD, UPDATE, or DELETE event).